plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,192 @@ func haveSame(a1, a2 []string) bool {
 	}
 	return false
 }
+
+type PodAffinityChecker struct {
+	info      NodeInfo
+	podLister algorithm.PodLister
+}
+
+func NewPodAffinityPredicate(info NodeInfo, podLister algorithm.PodLister) algorithm.FitPredicate {
+	checker := &PodAffinityChecker{
+		info:      info,
+		podLister: podLister,
+	}
+	return checker.InterPodAffinityMatches
+}
+
+func (checker *PodAffinityChecker) InterPodAffinityMatches(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
+	node, err := checker.info.GetNodeInfo(nodeName)
+	if err != nil {
+		return false, err
+	}
+	allPods, err := checker.podLister.List(labels.Everything())
+	if err != nil {
+		return false, err
+	}
+	return checker.NodeMatchPodAffinityAntiAffinity(pod, allPods, node), nil
+}
+
+// CheckIfAnyPodThatMatchPodAffinityTerm checks if any of given pods can match the specific podAffinityTerm.
+func (checker *PodAffinityChecker) CheckIfAnyPodThatMatchPodAffinityTerm(pod *api.Pod, allPods []*api.Pod, node *api.Node, podAffinityTerm api.PodAffinityTerm) (bool, error) {
+	for _, ep := range allPods {
+		match, err := priorityutil.CheckIfPodMatchPodAffinityTerm(ep, pod, podAffinityTerm,
+			func(ep *api.Pod) (*api.Node, error) { return checker.info.GetNodeInfo(ep.Spec.NodeName) },
+			func(pod *api.Pod) (*api.Node, error) { return node, nil },
+		)
+		if err != nil || match {
+			return match, err
+		}
+	}
+	return false, nil
+}
+
+// Checks whether the given node has pods which satisfy all the required pod affinity scheduling rules.
+// If node has pods which satisfy all the required pod affinity scheduling rules then return true.
+func (checker *PodAffinityChecker) CheckNodeMatchesHardPodAffinity(pod *api.Pod, allPods []*api.Pod, node *api.Node, podAffinity *api.PodAffinity) bool {
+	var podAffinityTerms []api.PodAffinityTerm
+	if len(podAffinity.RequiredDuringSchedulingIgnoredDuringExecution) != 0 {
+		podAffinityTerms = podAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+	}
+	// TODO: Uncomment this block when implement RequiredDuringSchedulingRequiredDuringExecution.
+	//if len(podAffinity.RequiredDuringSchedulingRequiredDuringExecution) != 0 {
+	//	podAffinityTerms = append(podAffinityTerms, podAffinity.RequiredDuringSchedulingRequiredDuringExecution...)
+	//}
+
+	for _, podAffinityTerm := range podAffinityTerms {
+		podAffinityTermMatches, err := checker.CheckIfAnyPodThatMatchPodAffinityTerm(pod, allPods, node, podAffinityTerm)
+		// if err or podAffinityTerm match failed then return false otherwise continue
```

Comment:
This comment isn't necessary. Also I think it is partly wrong anyway ("or podAffinityTerm match failed then return false" isn't correct since you return true even if podAffinityTerm match fails but you are in the "special case").